contracts/metatx/ERC2771Forwarder.sol

```diff
@@ -0,0 +1,278 @@
+     * case there's ETH left in the contract at the end of the execution instead of reverting. If
+     * the provided refund receiver is the `address(0)`, the contract will revert with an
+     * {ERC2771ForwarderMismatchedValue} error.
+     *
+     * NOTE: The `atomic` flag guarantees an all-or-nothing behavior only for the first level forwarded
+     * calls. In case a call is forwarded to another contract, it may revert without the top-level call
+     * reverting.
+     */
+    function executeBatch(
+        ForwardRequestData[] calldata requests,
+        bool atomic,
+        address payable refundReceiver
+    ) public payable virtual {
+        uint256 requestsValue;
+        uint256 spentValue;
+
+        for (uint256 i; i < requests.length; ++i) {
+            requestsValue += requests[i].value;
+            bool success = _execute(requests[i], _useNonce(requests[i].from), atomic);
+            if (success) {
+                // Will never reallistically overflow given _execute will natively revert at
+                // some point due to insufficient balance.
+                unchecked {
+                    spentValue += requests[i].value;
+                }
+            }
+        }
+
+        if (requestsValue != spentValue) {
+            if (refundReceiver == address(0)) {
+                revert ERC2771ForwarderMismatchedValue(requestsValue, msg.value);
+            }
+
+            // To avoid unexpected reverts because a request was frontrunned leaving ETH in the contract
+            // the value is sent back instead of reverting.
+            Address.sendValue(refundReceiver, requestsValue - spentValue);
+        }
+    }
+
+    /**
+     * @dev Validates if the provided request can be executed at current block with `request.signature`
+     * on behalf of `request.signer`.
+     *
+     * Internal function without nonce validation.
+     */
+    function _validate(
+        ForwardRequestData calldata request,
+        uint256 nonce
+    ) internal view virtual returns (bool alive, bool signerMatch, address signer) {
+        signer = _recoverForwardRequestSigner(request, nonce);
+        return (request.deadline >= block.timestamp, signer == request.from, signer);
+    }
+
+    /**
+     * @dev Recovers the signer of an EIP712 message hash for a forward `request` and its corresponding `signature`.
+     * See {ECDSA-recover}.
+     *
+     * Internal function without nonce validation.
+     */
+    function _recoverForwardRequestSigner(
+        ForwardRequestData calldata request,
+        uint256 nonce
+    ) internal view virtual returns (address) {
+        return
+            _hashTypedDataV4(
+                keccak256(
+                    abi.encode(
+                        _FORWARD_REQUEST_TYPEHASH,
+                        request.from,
+                        request.to,
+                        request.value,
+                        request.gas,
+                        nonce,
```

Comment:
I honestly think we should get `request.from`'s nounce here instead of passing it. 

Passing it might save one sload duplication when the nonce is consumed, but it also makes the function parameters more complex to follow IMO.

I'd value simplicity over a 100gas saving in an already super expensive operation.